app-catalog: enchance release management with filter and bulk action

Quent1L opened the pull request:
# Enhanced Release Management for App-Catalog Plugin

## Overview

This PR significantly improves the user experience for managing Helm releases in the app-catalog plugin by adding filtering, bulk operations, and quick actions directly from the list view.

## Key Features

### Smart Filtering
- **Name search**: Debounced text input for searching releases by name (300ms delay for optimal performance)
- **Namespace filter**: Autocomplete dropdown dynamically populated from available releases
- **Combined filtering**: Both filters work together for precise results
- **Consistent UX**: Filters positioned in the header, matching the Charts list design

### Bulk Delete Operations
- **Multi-selection**: Checkbox column for selecting individual or all releases
- **Selection toolbar**: Displays count of selected items with a delete button
- **Smart "Select All"**: Works with filtered results, supports indeterminate state
- **Confirmation dialog**: Prevents accidental bulk deletions with clear messaging
- **Parallel deletion**: Uses `Promise.all` for efficient bulk operations

### Quick Actions Menu
- **Inline actions**: Three-dot menu on each row for common operations
- **Upgrade**: Opens editor dialog without navigating to detail page
- **Rollback**: Shows version history with dates/times, automatically disabled for version 1
- **Delete**: Single-click access to delete confirmation
- **Consistent behavior**: Reuses all logic from the detail page for reliability

## User Impact

**Before**: Users had to navigate to each release's detail page to perform actions, with no way to filter or bulk manage releases. The rollback dialog only showed version numbers without timestamps.

**After**: Users can filter hundreds of releases instantly, perform bulk deletions, access all common actions with 2 clicks instead of 4+, and easily identify deployment versions by their timestamps when rolling back.

## Testing Considerations

- Verify filtering works correctly with special characters in names
- Test bulk delete with mix of successful/failed deletions
- Confirm rollback dialog shows correct version history
- Check "Select All" behavior with filters applied
- Validate timeout cleanup prevents memory leaks

## Screenshots

<img width="1456" height="334" alt="image" src="https://github.com/user-attachments/assets/c0587d86-569d-485b-9939-f922b452cb65" />

Quent1OK committed:
app-catalog: add filtering, bulk delete and quick-actions menu

Add name and namespace filter inputs to the release list header.
Add multi-select checkboxes for bulk uninstall with a confirmation
dialog. Add a per-row three-dot menu giving quick access to upgrade,
rollback and delete without navigating to the detail page.

Signed-off-by: Quent1L <quentin-ol@ik.me>

illume commented:
@Quent1L This is really awesome. Please give us some time to look through it :)

illume commented:
@Quent1L thanks for all those changes. I'll try it out in the next few days.

illume commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +145,335 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+          columns={[
+            {
+              label: '',
+              header: (
+                <Checkbox
+                  size="small"
+                  checked={
+                    filteredReleases
+                      ? filteredReleases.length > 0 &&
+                        selectedReleases.size === filteredReleases.length
+                      : false
+                  }
+                  indeterminate={
+                    filteredReleases
+                      ? selectedReleases.size > 0 && selectedReleases.size < filteredReleases.length
+                      : false
+                  }
+                  onChange={handleSelectAll}
+                />
```
A good suggestion.

Quent1L replied:
Fixed — aria-label="Select all releases" added to the Select All checkbox

illume commented on app-catalog/src/components/releases/RollbackDialog.tsx:
```diff
@@ -0,0 +1,73 @@
+        <Select
+          value={revertVersion}
+          onChange={event => onVersionChange(event.target.value as string)}
+          id="revert"
```
Seems like a good suggestion.

Quent1L replied:
Fixed

illume commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +173,443 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+        if (!parsed) return null;
+        return { namespace: parsed.namespace, name: parsed.name, key };
+      })
+      .filter((item): item is { namespace: string; name: string; key: string } => item !== null);
```
I guess it's fine as it is.

illume commented on app-catalog/src/components/releases/ReleaseActionsMenu.tsx:
```diff
@@ -0,0 +1,72 @@
+  const handleDelete = useCallback(() => {
+    onDelete(release);
+    handleMenuClose();
+  }, [release, onDelete, handleMenuClose]);
```
I guess this would be an improvement, but a bit nit picky.

illume commented on app-catalog/src/components/releases/RollbackDialog.tsx:
```diff
@@ -0,0 +1,90 @@
+            releaseHistory.releases.map((release: Release) => {
+              return (
+                <MenuItem key={release.version} value={release.version.toString()}>
+                  {release.version} - {new Date(release.info.last_deployed).toLocaleString()}
```
I don't think this is right. They are different views with different needs. I imagine here we are interested in the timestamp, because it's a rollback.

illume commented on app-catalog/src/components/releases/RollbackDialog.tsx:
```diff
@@ -0,0 +1,90 @@
+                  {release.version} - {new Date(release.info.last_deployed).toLocaleString()}
+                </MenuItem>
+              );
+            })}
```
This looks like a valid point.

illume commented on app-catalog/src/components/releases/RollbackDialog.tsx:
```diff
@@ -0,0 +1,90 @@
+          height: '100px',
+        }}
+      >
+        <InputLabel id="revert">Select a version</InputLabel>
+        <Select
+          value={revertVersion}
+          onChange={event => onVersionChange(event.target.value as string)}
+          id="revert"
+          labelId="revert"
+          fullWidth
+        >
+          {releaseHistory &&
+            releaseHistory.releases.map((release: Release) => {
+              return (
+                <MenuItem key={release.version} value={release.version.toString()}>
+                  {release.version} - {new Date(release.info.last_deployed).toLocaleString()}
+                </MenuItem>
+              );
+            })}
+        </Select>
```
This is a good suggestion.

illume replied:
@Quent1L this suggestion is still open

Quent1L replied:
Fixed — label="Select a version" added to the Select component.

illume reviewed (changes requested):
Thanks for this.

I went through all the copilot comments and marked a bunch of them as resolved when I thought they didn't make sense, were done already, or were too nitpicky.

There's some comments left open, can you please have a look?



The github CI check is failing with this typescript error:

```ts
npm run tsc

Error: src/components/releases/ReleaseList.stories.tsx(64,51): error TS2322: Type 'Promise<{ releases: { name: string; namespace: string; chart: { metadata: { name: string; appVersion: string; icon: string; }; }; version: string; info: { status: string; last_deployed: string; }; }[]; }>' is not assignable to type 'Promise<ReleasesResponse>'.
  Type '{ releases: { name: string; namespace: string; chart: { metadata: { name: string; appVersion: string; icon: string; }; }; version: string; info: { status: string; last_deployed: string; }; }[]; }' is not assignable to type 'ReleasesResponse'.
    Types of property 'releases' are incompatible.
      Type '{ name: string; namespace: string; chart: { metadata: { name: string; appVersion: string; icon: string; }; }; version: string; info: { status: string; last_deployed: string; }; }[]' is not assignable to type 'Release[]'.
        Type '{ name: string; namespace: string; chart: { metadata: { name: string; appVersion: string; icon: string; }; }; version: string; info: { status: string; last_deployed: string; }; }' is not assignable to type 'Release'.
          Types of property 'version' are incompatible.
            Type 'string' is not assignable to type 'number'.
Problem running tsc inside of "."
```

illume reviewed (changes requested):
Thanks for that!

I think it's in a good shape to get some more feedback from others. Can you please ask for someone else to review and test it in the headlamp Kubernetes slack channel?

btw. Can you please check the formatting? Seems the GitHub check is failing.
npm run format

Quent1L committed:
app-catalog: fix rollback to use numeric version and poll status

Pass the revision number as an integer to rollbackRelease instead of
a string. Add status polling after rollback so the list refreshes
once the operation completes, matching the delete flow.

Signed-off-by: Quent1L <quentin-ol@ik.me>

illume reviewed (changes requested):
@Quent1L great, thanks for all of this!

I went through and marked all the convos resolved that I could see were already addressed. Can you please resolve the remaining ones open if they've been addressed? (It's fine if you disagree with a comment, just mention why and mark it as resolved.)

I left some notes about a few things that would be good to document.

For the commits, can you please add app-catalog: and area to the parts changed. Generally it's easier to review if the commits are atomic in one area. See [Git commit guidelines in contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and git log for examples.

illume commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -288,42 +341,123 @@ export default function ReleaseList({ fetchReleases = listReleases }: ReleaseLis
     });
   }, [filteredReleases]);
 
+  const checkBulkDeleteComplete = useCallback(
```
This one needs documenting as well.

Quent1L replied:
Done — JSDoc added to checkBulkDeleteComplete describing the polling behavior and retry logic.

illume reviewed (changes requested):
<!-- prch:feedback -->
Thanks for these changes.

<!-- prch:commit-guidelines -->
Can you please have a look at the git commits to see if they meet the contribution guidelines? We use a Linux kernel style of git commits. See the [contributing guide](https://headlamp.dev/docs/latest/development/contributing/#commit-guidelines) for general context, and please see previous git commits with git log for examples.

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <description of changes>`.
- Keep the title and body lines under 72 characters.
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

<!-- prch:copilot-comments -->
Can you please address the open review comments?

Quent1L committed:
app-catalog: add JSDoc to new interfaces and functions

Document new props interfaces and key callbacks to satisfy the
project contribution guidelines. Remove accidentally committed
package-lock.json.

Signed-off-by: Quent1L <quentin-ol@ik.me>

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +145,335 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+              getter: (release: Release) => {
+                const key = `${release.namespace}/${release.name}`;
+                const isSelected = selectedReleases.has(key);
+                return (
+                  <Checkbox
+                    size="small"
+                    checked={isSelected}
+                    onChange={() => handleSelectRelease(release.name, release.namespace)}
+                  />
```
Fixed

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +173,443 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+          console.error('Failed to delete release:', error);
+          enqueueSnackbar(`Failed to delete release ${selectedRelease.name}`, {
+            variant: 'error',
+          });
```
Fixed — setIsDeleting(false) is called in the catch block.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+
+      // Check if any deleted releases still exist in the list
+      const stillExist = deletedKeys.some(key => {
+        const parsed = parseReleaseKey(key);
+        if (!parsed) return false;
+        const release = releases.find(
+          r => r.namespace === parsed.namespace && r.name === parsed.name
+        );
+        return release !== undefined;
+      });
+
+      if (stillExist) {
+        const outerTimeout = setTimeout(() => {
+          setRefetchCounter(prev => prev + 1); // Trigger fetch
+          const innerTimeout = setTimeout(() => {
+            checkBulkDeleteComplete(deletedKeys, retryCount + 1);
+          }, 500);
+          bulkDeleteTimeoutsRef.current.push(innerTimeout);
+        }, DELETE_STATUS_POLLING_INTERVAL);
+        bulkDeleteTimeoutsRef.current.push(outerTimeout);
```
Fixed — executed timeouts are removed from bulkDeleteTimeoutsRef.current via filter after they run.

Quent1L commented on app-catalog/src/components/releases/ReleaseFilters.tsx:
```diff
@@ -0,0 +1,64 @@
+  onNameFilterChange,
+  onNamespaceFilterChange,
+}: ReleaseFiltersProps) {
+  const [inputValue, setInputValue] = useState(nameFilter);
```
Fixed — useEffect(() => { setInputValue(nameFilter); }, [nameFilter]) added to sync local state.

Quent1L commented on app-catalog/src/components/releases/RollbackDialog.tsx:
```diff
@@ -0,0 +1,98 @@
+
+  return (
+    <Dialog open={open} maxWidth="xs" onClose={onCancel} title="Rollback">
+      <DialogContent
+        sx={{
+          width: { xs: '100%', sm: '400px' },
+          minHeight: '100px',
+        }}
```
The component uses sx props consistently throughout. The Detail.tsx approach used inline styles for a different layout context. Keeping sx here for consistency with the rest of the new components.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+            {
+              label: '',
+              header: (
+                <Checkbox
+                  size="small"
+                  checked={
+                    filteredReleases
+                      ? filteredReleases.length > 0 &&
+                        selectedReleases.size === filteredReleases.length
+                      : false
+                  }
+                  indeterminate={
+                    filteredReleases
+                      ? selectedReleases.size > 0 && selectedReleases.size < filteredReleases.length
+                      : false
+                  }
+                  onChange={handleSelectAll}
+                  aria-label="Select all releases"
+                />
```
Fixed — selectedVisibleCount counts only selected releases that are in filteredReleases, so the checked state correctly compares against the filtered view.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+            variant: 'info',
+          });
+          // Keep dialog open while polling - it will close on success in checkDeleteReleaseStatus
+          checkDeleteReleaseStatus(selectedRelease.name, selectedRelease.namespace);
```
Intentional design choice: keeping the dialog open with a loading state during polling gives users feedback that deletion is in progress. Detail.tsx closes immediately because it navigates away; in the list view, immediate

Quent1L committed:
app-catalog: fix delete race condition and clarify history entry type

 - Guard handleConfirmDelete against re-entry by setting isDeleting before the API call instead of after it resolves
 - Add isDeleting dep to handleConfirmDelete useCallback
 - Document stable useState setter refs omitted from handleConfirmBulkDelete deps
 - Rename Release -> ReleaseHistoryEntry in RollbackDialog to disambiguate from the full Release type in List

Signed-off-by: Quent1L <quentin-ol@ik.me>

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+
+      getActionStatus(name, 'uninstall')
+        .then(response => {
+          if (response.status === 'processing') {
+            deleteStatusTimeoutRef.current = setTimeout(
+              () => checkDeleteReleaseStatus(name, namespace, retryCount + 1),
+              DELETE_STATUS_POLLING_INTERVAL
+            );
+          } else if (response.status !== 'success') {
+            enqueueSnackbar(`Failed to delete release ${name}: ${response.message}`, {
+              variant: 'error',
+            });
+            setIsDeleting(false);
+            if (deleteStatusTimeoutRef.current) {
+              clearTimeout(deleteStatusTimeoutRef.current);
+              deleteStatusTimeoutRef.current = null;
+            }
+          } else {
+            enqueueSnackbar(`Successfully deleted release ${name}`, { variant: 'success' });
+            setOpenDeleteAlert(false);
+            setIsDeleting(false);
+            setRefetchCounter(prev => prev + 1);
+            if (deleteStatusTimeoutRef.current) {
+              clearTimeout(deleteStatusTimeoutRef.current);
+              deleteStatusTimeoutRef.current = null;
+            }
+          }
+        })
+        .catch(error => {
+          console.error('Failed to check delete status:', error);
+          enqueueSnackbar(`Failed to check delete status for ${name}`, { variant: 'error' });
+          setIsDeleting(false);
+          if (deleteStatusTimeoutRef.current) {
+            clearTimeout(deleteStatusTimeoutRef.current);
+            deleteStatusTimeoutRef.current = null;
+          }
+        });
+    },
+    [enqueueSnackbar, setRefetchCounter]
+  );
```
The cleanup effect clears deleteStatusTimeoutRef on unmount, stopping all future poll scheduling. In React 18, setState calls from in-flight promises on an unmounted component are silently ignored (no warnings, no errors). Adding an AbortController would add significant complexity for a case that is already benign at runtime.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
     ).then(entries => setLatestMap(Object.fromEntries(entries)));
   }, [releases]);
 
+  const filteredReleases = useMemo(() => {
+    if (!releases) return null;
+
+    return releases.filter(release => {
+      const matchesName =
+        !nameFilter || release.name.toLowerCase().includes(nameFilter.toLowerCase());
+      const matchesNamespace = !namespaceFilter || release.namespace === namespaceFilter;
+
+      return matchesName && matchesNamespace;
+    });
+  }, [releases, nameFilter, namespaceFilter]);
```
The `.toLowerCase()` call is idiomatic and has no measurable cost at typical cluster sizes. Pre-computing lowercase names would add a second useMemo and extra complexity without a demonstrated bottleneck. Happy to revisit if profiling shows this is slow at scale.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -63,66 +178,506 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+    }
+  }, [selectedRelease, enqueueSnackbar, checkDeleteReleaseStatus]);
+
+  const handleConfirmRollback = useCallback(() => {
```
Done — JSDoc added above handleConfirmRollback describing its behavior and side effects.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+        clearTimeout(deleteStatusTimeoutRef.current);
+        deleteStatusTimeoutRef.current = null;
+      }
+
+      deleteRelease(selectedRelease.namespace, selectedRelease.name)
+        .then(() => {
+          setIsDeleting(true);
+          enqueueSnackbar(`Delete request for release ${selectedRelease.name} accepted`, {
+            variant: 'info',
+          });
+          // Keep dialog open while polling - it will close on success in checkDeleteReleaseStatus
+          checkDeleteReleaseStatus(selectedRelease.name, selectedRelease.namespace);
+        })
+        .catch(error => {
+          console.error('Failed to delete release:', error);
+          enqueueSnackbar(`Failed to delete release ${selectedRelease.name}`, {
+            variant: 'error',
+          });
+          setIsDeleting(false);
+        });
+    }
+  }, [selectedRelease, enqueueSnackbar, checkDeleteReleaseStatus]);
+
+  const handleConfirmRollback = useCallback(() => {
+    if (selectedRelease) {
+      rollbackRelease(selectedRelease.namespace, selectedRelease.name, revertVersion)
+        .then(() => {
+          setRollbackPopup(false);
+          enqueueSnackbar(`Rollback successful for ${selectedRelease.name}`, {
+            variant: 'success',
+          });
+          setRefetchCounter(prev => prev + 1);
+        })
+        .catch(error => {
+          console.error('Failed to rollback release:', error);
+          enqueueSnackbar(`Failed to rollback ${selectedRelease.name}`, {
+            variant: 'error',
+          });
+        });
+    }
```
The catch block closes the dialog and shows a detailed error snackbar (Failed to rollback ${releaseName}), matching the second option you described ("closing it while showing a detailed error snackbar"). The user gets  immediate visual feedback via the snackbar

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +173,443 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+          .map(item => item.key);
+
+        if (failed === 0) {
+          enqueueSnackbar(
+            `Successfully initiated deletion of ${succeeded} release(s)`,
+            { variant: 'info' }
+          );
+        } else if (succeeded === 0) {
+          enqueueSnackbar(`Failed to delete all ${failed} release(s)`, { variant: 'error' });
+          setIsBulkDeleting(false);
+          setOpenBulkDeleteAlert(false);
+          setSelectedReleases(new Set());
+          return;
+        } else {
+          enqueueSnackbar(
+            `Initiated deletion of ${succeeded} release(s), failed ${failed}`,
+            { variant: 'warning' }
+          );
+        }
+
+        setOpenBulkDeleteAlert(false);
+        setSelectedReleases(new Set());
+
+        if (successfullyDeletedKeys.length > 0) {
+          setUpdate(prev => !prev);
+          setTimeout(
+            () => checkBulkDeleteComplete(successfullyDeletedKeys),
+            DELETE_STATUS_POLLING_INTERVAL
+          );
+        } else {
+          setIsBulkDeleting(false);
+        }
+      })
+      .catch(error => {
+        console.error('Unexpected error in bulk delete:', error);
+        enqueueSnackbar('Unexpected error during bulk deletion', { variant: 'error' });
+        setIsBulkDeleting(false);
+        setOpenBulkDeleteAlert(false);
+      });
+  }, [selectedReleases, enqueueSnackbar, checkBulkDeleteComplete]);
```
Addressed — `bulkDeleteTimeoutsRef` tracks all active timeouts across bulk delete operations. Each completed timeout removes itself from the ref array, and the cleanup effect clears all remaining timeouts on unmount. A concurrent bulk delete would share the same ref and the final `setIsBulkDeleting(false)` is only called once all deleted releases are confirmed gone.

Quent1L commented on app-catalog/src/components/releases/List.tsx:
```diff
@@ -61,66 +176,446 @@ export default function ReleaseList({ fetchReleases = listReleases }) {
+        setIsBulkDeleting(false);
+        setOpenBulkDeleteAlert(false);
+      });
+  }, [selectedReleases, enqueueSnackbar, checkBulkDeleteComplete]);
```
State setters from useState (setIsBulkDeleting, setOpenBulkDeleteAlert, setSelectedReleases, setRefetchCounter) are guaranteed stable refs per the React docs — they never change between renders. Added a comment in the code documenting this intention explicitly.

Quent1L commented on app-catalog/src/components/releases/ReleaseActionsMenu.tsx:
```diff
@@ -0,0 +1,72 @@
+
+const ICON_SPACING = { marginRight: 8 };
+
+interface Release {
```
Done — JSDoc added to both `Release` and `ReleaseActionsMenuProps` interfaces.

Quent1L commented on app-catalog/src/components/releases/ReleaseFilters.tsx:
```diff
@@ -0,0 +1,68 @@
+import { Autocomplete, Box, TextField } from '@mui/material';
+import { useEffect, useRef, useState } from 'react';
+
+interface ReleaseFiltersProps {
```
Done — JSDoc added to ReleaseFiltersProps with field-level descriptions.

Quent1L commented:
I believe I've addressed all the review comments. Apologies for the delay in responding,  I wasn't very available. Let me know what you think @illume !

yolossn commented:
Tested this and works as expected. There are few minor changes required but we will take it from here.
